README.md
# World Edit

Rising World editing tool.


## Introduction

This script is based on the original World Edit script by andyzee. The main difference with the original script are the commands.

Instead of `/we-select`, players must type `/we select` (no dash). Also, `/we-fillblock` has been replaced with `/we place` with extended arguments parameters.


## Installation

This modules has dependencies, therefore you need to make sure that all of them are properly installed before using this script!

### Using Git

Go to the `scripts` folder of your Rising World installation and type

```
git clone --recursive https://github.com/RisingWorld/world-edit.git
```

### Manually

Download the Zip file for [this](https://github.com/RisingWorld/world-edit/archive/master.zip) repository and extract it to your Rising World's `scripts/world-edit` folder.

Download the Zip file for the [i18n](https://github.com/RisingWorld/i18n/archive/master.zip) sub-module, and extract it inside the `i18n` folder of this script.

Download the Zip file for the [command parser](https://github.com/RisingWorld/command-parser/archive/master.zip) sub-module, and extract it inside the `command-parser` folder of this script.

Download the Zip file for the [lua-ext](https://github.com/RisingWorld/lua-ext/archive/master.zip) sub-module, and extract it inside the `lua-ext` folder of this script.

Your final script folder should look somewhat like this

```
./risingworld/scripts/world-edit
   ./command-parser
      ./parse-args.lua
   ./i18n
      ./i18n.lua
   ./lc_messages
      ./en.locale
      ./de.locale
   ./listeners
      ./commandListener.lua
      ./playerListener.lua
   ./lua-ext
      ./string-ext.lua
      ./table-ext.lua
   ./definition.xml
   ./blocks.lua
   ./config.properties
   ./security.lua
   ./worldedit.lua
```

(**Note**: there are more files, but only the necessary ones are shown, here.)


## Updating

Whenever world-edit is updated, you should also update your server. To keep up-to-date with the newest features, but more importantly to stay up-to-date with the most recent patches of Rising World, and correct any security issues. You may also consider automating this process. The updates will take effect only after server restart.

### Using Git

Go to your `world-edit` script folder and type

```
git fetch --recurse-submodules origin master
```

### Manually

Repeat manual installation process, overwrite any existing files.


## Usage

In-game, in chat, type `/we <command>` where `<command>` is one of the following :

### Commands

* `help [command]` : dipslay help. If `command` is specified, display help for that command.  

  Ex: `/we help fill`

* `select` : start area selection  
* `cancel` : cancel area selection
* `fill <texture|#id> [-c]` : fill the selected area with the specified terrain.  Add `-c` to clear everything, first. All available textures are 
  
  Ex: `/we fill -c grass`

* `clear [obj|con|veg|block|all|abs]` : clear the selected area of (obj)ects, (con)structions, (veg)etations, (block)s, (all), or (abs)olutely everything. (Default `all`)  

  Ex: `/we clear veg`

* `paint <start|stop> <texture> [size] [height]` : paint the terrain as the player walks using the specified terrain and brush size (defaults to `1`) and height adjustement (defaults to `0`). (**Note**: the brush is square at the moment. And the terrain can only be raised.) Use `/we paint stop` to stop. It is possible to call `/we paint start ...` multiple times in a row, changing texture and brush size at will.
  
  Ex: `/we paint start 6 3` ... `/we paint stop`

  Ex: `/we paint start 1` ... `/we paint stop`

  Ex: `/we paint start 9 2 1` ... `/we paint stop`

* `place <blockType> <id> [north|east|south|west [sideway|flipped]]` : place a block with the given `id`, optionally facing the given direction and put `sideway` or `flipped`.  

  Ex: `/we place ramp 121 east flipped`

* `plant <areatype> <plants>` : plant vegetations (trees, flowers, etc.). 
  * `<areatype>` may be one of the following :
    * `single` plant a single plant right in front of the character.
    * `line <distance> <count>` plant in a straight line to the nearest axis, up to `distance` blocks. The value `count` may be an absolute value or a percentage of the total `distance`.
    * `freeline <distance> <count>` plant in the direction the character is facing, where `distance` and `count` are the same as for the above.
    * `rect <ns> <ew> <count> [-b]` plant in a rectangle **around** the character (character at the very center of the area). Both `ns` and `ew` are mandatory and determine the total width of each side. The amount of vegetation planted is determined by `count` and may be an absolute value or a percentage of the rectangle area. Use the optional `-b` (border) flag to plant at the edge of the area only, and `count` will be the perimeter instead of the area.
    * `circle <diameter> <count> [-b]` plant in a circle with the specified `diameter` around the character (character at the very center of the area). The amount of vegetation planted is determined by `count` and may be an absolute value or a percentage of the circle area. Use the `-b` (border) flag to plant at the edge of the area only, and `count` will be based on the circonference instead of the area.
  * `<plants>` is a space delimited list of plants ids to plant. To specify a range of plants, use `..` between two numbers (no space). For example `1 2 3 4 7 9 10 11 12` is the same as `1..4 7 9..12`.

  Ex: `/we plant single 13..27` plant a single random flower

  Ex: `/we plant line 10 7 29` plant 7 trees at random on a single straight line of 10 blocks long, on the nearest axis, in front of the character

  Ex: `/we plant freeline 10 50% 6 -e` plant 5 pine trees (50% of 10) spaced evenly on a single straight line, 10 blocks long, facing and in front of the character.

  Ex: `/we plant rect 30 10 100 2..4 6 7 -b` plant 100 trees at random around a perimeter of a rectangle of 30 blocks (north-south) by 10 blocks (east-west)

  Ex: `/we plant circle 20 30% 2 3 4` plant maple trees in a circle area with a diameter of 20 blocks filling it at 30%


### Textures

+ air (id `0`)
+ dirt (id `1`)
+ grass (id `2`)
+ stone (id `3`)
+ gravel (id `4`)
+ rock (id `5`)
+ farmland (id `6`)
+ mud (id `7`)
+ snow (id `8`)
+ sand (id `9`)
+ desertdirt (id `10`)
+ desertstone (id `11`)
+ clay (id `12`)
+ dungeonwall (id `13`)
+ dungeonfloor (id `14`)
+ bonewall (id `15`)
+ hellstone (id `16`)
+ iron (id `-101`)
+ copper (id `-102`)
+ aluminium (id `-103`)
+ silver (id `-104`)
+ gold (id `-105`)
+ tungsten (id `-106`)
+ cobalt (id `-107`)
+ mithril (id `-108`)
+ grass9 (id `-10`)
+ grass8 (id `-9`)
+ grass7 (id `-8`)
+ grass6 (id `-7`)
+ grass5 (id `-6`)
+ grass4 (id `-5`)
+ grass3 (id `-4`)
+ grass2 (id `-3`)
+ grass1 (id `-2`)


### Block Types

* `block` (aliases: `b`, `blk`)
* `cylinder` (aliases: `c`, `cyl`)
* `cylinderhalf` (aliases: `ch`, `cylh`)
* `stair` (aliases: `s`, `s1`, `stair1`)
* `stair2` (aliases: `s2`)
* `stair3` (aliases: `s3`)
* `staircorner` (aliases: `sc`, `stairc`)
* `stairinnercorner` (aliases: `sic`, `stairic`)
* `ramp` (aliases: `r`)
* `ramphalfcorner` (aliases: `rhc`, `ramphc`, `ramphalfc`)
* `rampinnercorner` (aliases: `ric`, `rampic`)
* `rampcorner` (aliases: `rc`, `rampc`)
* `halfblockbottom` (aliases: `hb`, `hb1`, `hbb`, `halfblk`, `halfblk1`, `halfblkb`, `halfblock`, `halfblock1` `halfblockb`)
* `halfblockcenter` (aliases: `hb2`, `hbc`, `halfblk2`, `halfblkc`, `halfblock2`, `halfblockc`)
* `halfblocktop` (aliases: `hb3`, `hbt`, `halfblk3`, `halfblkt`, `halfblock3`, `halfblockt`)
* `pyramid` (aliases: `p`, `pyr`)
* `arc` (aliases: `a`)


### Plants

+ scrub1 (id `1`)
+ mapletree1 (id `2`)
+ mapletree2 (id `3`)
+ mapletree3 (id `4`)
+ dead1 (id `5`)
+ pine1 (id `6`)
+ forestpine (id `7`)
+ flower1_s (id `8`)
+ flower1_m (id `9`)
+ flower1_l (id `10`)
+ fern1 (id `11`)
+ pumpkin1 (id `12`)
+ flower2_s (id `13`)
+ flower2_m (id `14`)
+ flower2_l (id `15`)
+ flower3_s (id `16`)
+ flower3_m (id `17`)
+ flower3_l (id `18`)
+ flower4_s (id `19`)
+ flower4_m (id `20`)
+ flower4_l (id `21`)
+ flower5_s (id `22`)
+ flower5_m (id `23`)
+ flower5_l (id `24`)
+ flower6_s (id `25`)
+ flower6_m (id `26`)
+ flower6_l (id `27`)
+ maplesapling_s0 (id `28`)
+ maplesapling_s1 (id `29`)
+ pinesapling_s0 (id `30`)
+ pinesapling_s1 (id `31`)
+ watermelon1 (id `32`)
+ watermelon2 (id `33`)
+ tomato_s0 (id `34`)
+ tomato_s1 (id `35`)
+ tomato_s2 (id `36`)
+ tomato_s3 (id `37`)
+ tomato (id `38`)
+ carrot_s0 (id `39`)
+ carrot_s1 (id `40`)
+ carrot (id `41`)


## Contributors

* LordFoobar (Yanick Rochon)
* Yahgiggle (Deon Hamilton)
* andyzee (Andy Zee)

### Translators

* NDMR (German)


## License

Copyright (c) 2015 Rising World Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
## 2015-06-15

* Fixed minor issues with /we plant line
* Cleaned some code (removed useless commented code)
* Implemented /we paint

## 2015-09-13

* Implemented `/we plant`
* Moved `string-ext` and `table-ext` to a combined `lua-ext` for less module dependencies

## 2015-05-12

* Added `string-ext` sub-module to wrap long lines when displaying help (i.e. `/we help`)
* All terrain textures are now listed and available with `/we fill`, type `/we help fill` for more info.

## 2015-05-01

* Moved `table-ext.lua` in it's own [sub-module](https://github.com/RisingWorld/table-ext).
* the `config` object is no longer a global variable
* Updated README

## 2015-04-28

* fixed label not being reset on `/we cancel` if no selection was made. This was resulted after the update 0.5.6.7, which fixed possible invalid selection in the marking selector API.
* All commands involving a marking selector now have a `-p` (i.e. "preserve") flag to preserve the current marking selector status object (if one exists).

blocks.lua

--- Thank you yahgiggle !!
--- https://forum.rising-world.net/index.php/Thread/1210-SCRIPT-World-Edit/?postID=12454#post12454
local blockDef = {
  block = { north = 0, east = 0,  south = 0, west = 0 },

  cylinder = { north = 3, east = 3,  south = 3, west = 3 },
  cylindersideway = { north = 204, east = 207,  south = 204, west = 207 },

  cylinderhalf = { north = 6, east = 9,  south = 12, west = 15 },

  stair = {  north = 39, east = 42,  south = 45, west = 48 },
  stairsideway = { north = 273, east = 276,  south = 279, west = 282 },
  stairflipped = { north = 285, east = 288,  south = 291, west = 294 },

  stair2 = { north = 51, east = 54,  south = 57, west = 60 },

  stair3 = { north = 63, east = 66,  south = 69, west = 72 },

  staircorner = { north = 75, east = 78,  south = 81, west = 84 },
  staircornerflipped = { north = 297, east = 300,  south = 303, west = 306 },

  stairinnercorner = { north = 87, east = 90,  south = 93, west = 96 },
  stairinnercornerflipped = { north = 309, east = 312,  south = 315, west = 318 },

  ramp = { north = 99, east = 102,  south = 105, west = 108 },
  rampsideway = { north = 225, east = 228,  south = 231, west = 234 },
  rampflipped = { north = 111, east = 114,  south = 117, west = 120 },

  ramphalfcorner = { north = 123, east = 126,  south = 129, west = 132 },
  ramphalfcornerflipped = { north = 147, east = 150,  south = 153, west = 156 },

  rampinnercorner = { north = 135, east = 138,  south = 141, west = 144 },
  rampinnercornerflipped = { north = 159, east = 162,  south = 165, west = 168 },

  rampcorner = { north = 171, east = 174,  south = 177, west = 180 },
  rampcornerflipped = { north = 183, east = 186,  south = 189, west = 192 },

  halfblockbottom = { north = 195, east = 195,  south = 195, west = 195 },
  halfblockcenter = { north = 198, east = 198,  south = 198, west = 198 },
  halfblocktop = { north = 201, east = 201,  south = 201, west = 201 },

  pyramid = { north = 219, east = 219,  south = 219, west = 219 },
  pyramidflipped = { north = 222, east = 222,  south = 222, west = 222 },

  arc = { north = 237, east = 240,  south = 243, west = 246 },
  arcflipped = { north = 249, east = 252,  south = 255, west = 258 },
  arcsideway = { north = 261, east = 264,  south = 267, west = 270 }
};
local directionOrder = { "north", "east", "south", "west" };
local UNKNOWN_BLOCK_ID = 99999999;

--- Add any alias blocks may have to the blockDef table
local blockAliasMap = {
  b = "block",
  blk = "block",

  c = "cylinder",
  cyl = "cylinder",

  ch = "cylinderhalf",
  cylh = "cylinderhalf",

  s = "stair",
  s1 = "stair",
  stair1 = "stair",

  s2 = "stair2",

  s3 = "stair3",

  sc = "staircorner",
  stairc = "staircorner",

  sic = "stairinnercorner",
  stairic = "stairinnercorner",

  r = "ramp",

  rhc = "ramphalfcorner",
  ramphc = "ramphalfcorner",
  ramphalfc = "ramphalfcorner",

  ric = "rampinnercorner",
  rampic = "rampinnercorner",

  rc = "rampcorner",
  rampc = "rampcorner",

  hb = "halfblockbottom",
  hbb = "halfblockbottom",
  hb1 = "halfblockbottom",
  halfblk = "halfblockbottom",
  halfblkb = "halfblockbottom",
  halfblk1 = "halfblockbottom",
  halfblock = "halfblockbottom",
  halfblockb = "halfblockbottom",
  halfblock1 = "halfblockbottom",
  hbc = "halfblockcenter",
  hb2 = "halfblockcenter",
  halfblkc = "halfblockcenter",
  halfblockc = "halfblockcenter",
  halfblock2 = "halfblockcenter",
  hbt = "halfblocktop",
  hb3 = "halfblocktop",
  halfblkt = "halfblocktop",
  halfblockt = "halfblocktop",
  halfblock3 = "halfblocktop",

  p = "pyramid",
  pyr = "pyramid",

  a = "arc"
};


--- Return the absolute block id for the given variation, type, orientation and direction
-- @param id the variation texture id of the block
-- @param blockType a block type as string (see aliasses)
-- @param direction may be "north", "south", "east", or "west"
-- @param orientation may be "", "sideway" or "flipped"
-- @return a numeric id
function getBlockId(id, blockType, direction, orientation)
  local idDef;

  if id == nil then
    return nil;
  end;

  blockType = blockType and blockAliasMap[blockType] or blockType;
  direction = direction and direction or "north";
  orientation = orientation and orientation or "";

  idDef = blockDef[blockType..orientation] or blockDef[blockType];

  if idDef and idDef[direction] then
    id = (idDef[direction] * 100) + id;
  end

  return id;
end


--- Return all valid block types
-- @return table
function getBlockTypes()
  local types = {};
  for key,val in pairs(blockDef) do
    if string.sub(key, -7) ~= "flipped" and string.sub(key, -7) ~= "sideway" then
      table.insert(types, key);
    end
  end

  table.sort(types);

  return types;
end


--- Reverse function to return the block id, type, and orientation from
--- an absolute block id
-- @param absId the absolute block id
-- @return table
function getBlockTypeAndId(absId)
  local id = UNKNOWN_BLOCK_ID;
  local blockType = "block";
  local direction = "north";
  local orientation = nil;

  for type,def in pairs(blockDef) do
    for k,dir in next, directionOrder do
      local baseId = def[dir];

      if absId > (baseId*100) and id > math.abs(absId - (baseId*100)) then
        id = absId - (baseId * 100);
        direction = dir;

        if string.sub(type, -7) == "flipped" or string.sub(type, -7) == "sideway" then
          blockType = string.sub(type, 0, -8);
          orientation = string.sub(type, -7);
        else
          blockType = type;
          orientation = nil;
        end
      end
    end
  end

  if id == UNKNOWN_BLOCK_ID then
    id = absId;
  end

  return {
    absId = absId,
    id = id,
    type = blockType,
    direction = direction,
    orientation = orientation
  };
end


listeners/commandListener.lua

include("command-parser/parse-args.lua");

-- set maximum number of plants to be planted at once
local MAX_PLANTS_COUNT = 500;

-- set maximum distance plants can be planted from character
local MAX_PLANTS_DISTANCE = 64;


local textureAliasMap = {
  grass9 = -10,
  grass8 = -9,
  grass7 = -8,
  grass6 = -7,
  grass5 = -6,
  grass4 = -5,
  grass3 = -4,
  grass2 = -3,
  grass1 = -2,

  air = 0,
  dirt = 1,
  grass = 2,
  stone = 3,
  gravel = 4,
  rock = 5,
  farmland = 6,
  mud = 7,
  snow = 8,
  sand = 9,
  desertdirt = 10,
  desertstone = 11,
  clay = 12,
  dungeonwall = 13,
  dungeonfloor = 14,
  bonewall = 15,
  hellstone = 16,
  iron = -101,
  copper = -102,
  aluminium = -103,
  silver = -104,
  gold = -105,
  tungsten = -106,
  cobalt = -107,
  mithril = -108
};
local fillAvailableArgs = table.keys(textureAliasMap);
local clearAvailableArgs = {"obj","con","veg","block","all","abs"};

table.sort(fillAvailableArgs, function (a, b)
  return textureAliasMap[a] < textureAliasMap[b];
end);
table.insert(fillAvailableArgs, "#id");


local function setLabel(event, text)
  local label = event.player:getAttribute("weStateLabel");

  if text then
    label:setText(text);
    label:setVisible(true);
  else
    label:setText("");
    label:setVisible(false);
  end
end


local function weAbout(event)
  event.player:sendTextMessage("[#FFFF00]World-Edit "..VERSION);
end


local function weHelp(event, args)
  local helpContext = string.lower(args[1] or "");

  if helpContext == "select" then
    event.player:sendTextMessage("[#33FF33]/we select");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.select.usage"));
  elseif helpContext == "cancel" then
    event.player:sendTextMessage("[#33FF33]/we cancel");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.cancel.usage"));
  elseif helpContext == "clear" then
    event.player:sendTextMessage("[#33FF33]/we clear ["..table.concat(clearAvailableArgs, '|').."] [-p]");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.clear.usage"));
  elseif helpContext == "fill" then
    event.player:sendTextMessage("[#33FF33]/we fill <texture> [-p]");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.fill.usage", table.concat(fillAvailableArgs, ', ')));
  elseif helpContext == "paint" then
    event.player:sendTextMessage("[#33FF33]/we paint <start|stop> <texture> [size] [height]");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.paint.usage", table.concat(fillAvailableArgs, ', ')));
  elseif helpContext == "place" then
    event.player:sendTextMessage("[#33FF33]/we place <blocktype> <id> [north|east|south|west] [sideway|flipped] [-p]");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.place.usage", "blocktype", table.concat(getBlockTypes(), ', ')));
  elseif helpContext == "plant" then
    event.player:sendTextMessage("[#33FF33]/we plant <areatype> <ids>");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.plant.usage"));
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.plant.example"));
  elseif helpContext == "about" then
    weAbout(event);
  else
    event.player:sendTextMessage("[#33FF33]/we <about|help|select|cancel|clear|fill|paint|place|plant> [args]");
    event.player:sendTextMessage("[#FFFF00]"..i18n.t(event.player, "help.usage", "/we help fill"));
  end
end


local function weSelect(event)
  event.player:enableMarkingSelector(function()
    setLabel(event, i18n.t(event.player, "select.start"));
  end);
end


local function weCancel(event)
  event.player:disableMarkingSelector(function(markingEvent)
    setLabel(event);

    if markingEvent ~= false then
      event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "select.cancelled"));
    end
  end);
end


local function weClear(event, args, flags)
  local clearObjType = string.lower(args[1] or "all");

  event.player:getMarkingSelectorStatus(function(markingEvent)
    if markingEvent ~= false then
      if clearObjType == "obj" then
        --print("Clearing area of objects");
        removeObjects(markingEvent);
      elseif clearObjType == "con" then
        --print("Clearing area of construction");
        removeConstr(markingEvent);
      elseif clearObjType == "veg" then
        --print("Clearing area of vegetation");
        removeVeg(markingEvent);
      elseif clearObjType == "block" then
        --print("Clearing area of Blocks");
        fillBlock(markingEvent, 0);
      elseif clearObjType == "all" then
        --print("Clearing area of all");
        removeAll(markingEvent, false);
      elseif clearObjType == "abs" then
        --print("Clearing area of absolutely everything");
        removeAll(markingEvent, true);
      else
        return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.use.args", table.concat(clearAvailableArgs, ", ")));
      end

      if not flags["p"] then
        event.player:disableMarkingSelector(function()
          setLabel(event);
        end);
      end
    else
      event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.no.selection"));
    end
  end);

end


local function wePlaceBlock(event, args, flags)
  local blockType = args[1] and string.lower(args[1]);
  local direction = args[3] and string.lower(args[3]) or "north";
  local orientation = args[4] and string.lower(args[4]) or "";
  local blockId = args[2] and getBlockId(tonumber(args[2]) or 0, blockType, direction, orientation);

  if blockType and blockId then

    if blockId < 21 and blockId ~= 0 then
      --print("Block id adjusted from "..blockId.." to 21");
      blockId = 21;
    end;

    event.player:getMarkingSelectorStatus(function(markingEvent)
      if markingEvent ~= false then
        --print("Placing "..blockType.." in area with id "..blockId..(cleanup ~= nil and " with cleanup" or ""));
        fillBlock(markingEvent, blockId);

        if not flags["p"] then
          event.player:disableMarkingSelector(function()
            setLabel(event);
          end);
        end
      else
        event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.no.selection"));
      end
    end);
  elseif blockType == nil then
    event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.missing.arg", "type"));
  elseif blockId == nil then
    event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.missing.arg", "id"));
  end
end


local function wePlant(event, args, flags)
  local precision = 10000;
  local ids = {};
  local total;
  local count;
  local originPos = event.player:getPosition();
  local getPosition;
  local plantAngle;
  local id, pos;

  local function _scanIds(argIndex)
    -- build ids table of all the possible ids to use
    for i = argIndex, #args do
      local id = args[i];

      -- TODO : check for alias range

      local a, b = string.match(id, "(%d+)..(%d+)");

      if a and b then
        a = math.max(tonumber(a) or 1, 1);
        b = math.min(tonumber(b) or 1, 41);

        for i = a, b do
          table.insert(ids, i);
        end
      else
        id = tonumber(id);

        if id == nil then
          return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.invalid.arg", "range"));
        end

        table.insert(ids, math.max(math.min(id, 41), 1));
      end
    end
  end

  local function _count(argIndex, total)
    local count, percent = string.match(args[argIndex], "(%d+)(%%?)");

    if count == nil then
      return nil;
    elseif percent == "%" then
      count = (count / 100) * total;
    end

    return math.min(math.ceil(count), MAX_PLANTS_COUNT);
  end


  if args[1] == "single" then
    count = 1;
    getPosition = function ()
      local localPos = originPos:add(event.player:getViewDirection():setY(0):mult(1.5));
      return localPos.x, localPos.y - 2, localPos.z;
    end;

    _scanIds(2);
  elseif args[1] == "line" or args[1] == "freeline" then
    local distance = math.min(tonumber(args[2]) or 1, MAX_PLANTS_DISTANCE);
    local direction = event.player:getViewDirection():clone():setY(0);

    count = _count(3, distance);

    if flags["e"] then
      local increment = distance / count;
      local step = increment / 2;

      --print("COUNT / DISTANCE = INCREMENT = ".. count .. " / " .. distance .. " = " .. increment);
      getPosition = function ()
        local localPos = originPos:add(direction:mult(step));

        step = step + increment;

        return localPos.x, localPos.y - 2, localPos.z;
      end;
    else
      getPosition = function ()
        local localPos = originPos:add(direction:mult(math.random(1, distance * precision) / precision));
        return localPos.x, localPos.y - 2, localPos.z;
      end;
    end

    -- lock direction to nearest axis
    if args[1] == "line" then
      if math.abs(direction.x) > math.abs(direction.z) then
        if direction.x < 0 then
          direction.x = -1;
        else
          direction.x = 1;
        end
        direction.z = 0;
      else
        direction.x = 0;
        if direction.z < 0 then
          direction.z = -1;
        else
          direction.z = 1;
        end
      end
    end
    _scanIds(4);
  elseif args[1] == "rect" then
    local sizeZ = math.min(tonumber(args[2]) or 1, MAX_PLANTS_DISTANCE);  -- north south
    local sizeX = math.min(tonumber(args[3]) or 1, MAX_PLANTS_DISTANCE);  -- east west

    if flags["b"] then
      count = _count(4, (sizeX * 2) + (sizeZ * 2));
      getPosition = function ()
        local x;
        local y = originPos.y - 2;
        local z;
        local s = math.random(0, 19) % 4;

        if s == 0 or s == 2 then
          x = originPos.x + math.random(-math.floor(sizeX / 2) * precision, math.ceil(sizeX / 2) * precision) / precision;
          if s == 0 then
            z = originPos.z - sizeZ;
          else
            z = originPos.z + sizeZ;
          end
        else
          if s == 1 then
            x = originPos.x - sizeX;
          else
            x = originPos.x + sizeX;
          end
          z = originPos.z + math.random(-math.floor(sizeZ / 2) * precision, math.ceil(sizeZ / 2) * precision) / precision;
        end

        return x, y, z;
      end;
    else
      count = _count(4, sizeX * sizeZ);
      getPosition = function ()
        local x = originPos.x + math.random(-math.floor(sizeX / 2) * precision, math.ceil(sizeX / 2) * precision) / precision;
        local y = originPos.y - 2;
        local z = originPos.z + math.random(-math.floor(sizeZ / 2) * precision, math.ceil(sizeZ / 2) * precision) / precision;

        return x, y, z;
      end;
    end
    _scanIds(5);
  elseif args[1] == "circle" then
    local radius = math.min((tonumber(args[2]) or 1) / 2, MAX_PLANTS_DISTANCE);

    if flags["b"] then
      count = _count(3, math.pi * radius * 2);
      getPosition = function ()
        local angle = (math.random(0, 359 * precision) / precision) * math.pi / 180;
        local x = originPos.x + (radius * math.cos(angle));
        local y = originPos.y - 2;
        local z = originPos.z + (radius * math.sin(angle));

        return x, y, z;
      end;
    else
      count = _count(3, math.pi * radius * radius);
      getPosition = function ()
        local angle = (math.random(0, 359 * precision) / precision) * math.pi / 180;
        local x = originPos.x + ((math.random(0, radius * precision) / precision) * math.cos(angle));
        local y = originPos.y - 2;
        local z = originPos.z + ((math.random(0, radius * precision) / precision) * math.sin(angle));

        return x, y, z;
      end;
    end

    _scanIds(4);
  else
    return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.invalid.arg", "areatype"));
  end

  if #ids == 0 then
    return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.missing.arg", "plants"));
  elseif count == nil then
    return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.invalid.arg", "count"));
  end

  for i = 1, count do
    id = ids[math.random(1, #ids)];
    plantAngle = math.random(0, 359 * precision) / precision;
    pos = findNearestTerrainFloor(getPosition());

    --print("Planting #" .. i .. " / " .. count .. ": " .. id .. " @ " .. pos.x .. "," .. pos.y .. "," .. pos.z .. " : " .. plantAngle);

    placeVegetation(pos.x, pos.y - (0.2 + (math.random(0, 20) / 100)), pos.z, plantAngle, id);
  end

end


local function weFill(event, args, flags)
  local cleanup = flags["c"] or flags["clean"];
  local id = tonumber(args[1]) or textureAliasMap[string.lower(args[1] or "")];

  if id then

    if not table.contains(textureAliasMap, id) then
      --print("Terrain id adjusted from "..id.." to "..textureAliasMap["air"]);
      id = textureAliasMap["air"];
    end;

    event.player:getMarkingSelectorStatus(function(markingEvent)
      if markingEvent ~= false then
        --print("Filling area with id "..id..(cleanup ~= nil and " with cleanup" or ""));
        if cleanup then
          removeAll(markingEvent);
        end

        fillTerrain(markingEvent, id);

        if not flags["p"] then
          event.player:disableMarkingSelector(function()
            setLabel(event);
          end);
        end
      else
        event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.no.selection"));
      end
    end);
  else
    event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.use.args", table.concat(fillAvailableArgs, ', ')));
  end
end


local function wePaint(event, args, flags)
  local cmd = args[1];

  if cmd == "start" then
    local id = tonumber(args[2]) or textureAliasMap[string.lower(args[2] or "")];
    local brushSize = math.min(math.max(tonumber(args[3]) or 1, 1), 10);
    local heightAdjust = math.min(math.max(tonumber(args[4]) or 0, 0), 16);

    if id then
      if id == 0 then
        return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.invalid.arg", "texture"));
      end

      local textureName = id;
      for key,texture in pairs(textureAliasMap) do
        if texture == id then
          textureName = key;
        end
      end

      setLabel(event, i18n.t(event.player, "paint.start", textureName));
      event.player:setAttribute("painting", {
        id = id,
        brushSize = brushSize,
        height = heightAdjust
      });
    else
      return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.use.args", table.concat(fillAvailableArgs, ', ')));
    end
  elseif cmd == "stop" then
    setLabel(event);
    event.player:setAttribute("painting", nil);
    event.player:setAttribute("painting.brush.size", nil);
  else
    return event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.missing.arg", "start|stop"));
  end
end

local function wePaintProcess(event)
  local paint = event.player:getAttribute("painting");
  local playerPos = event.player:getPosition();
  local pos, terrainId = findNearestTerrainFloor(playerPos.x, playerPos.y, playerPos.z);

  if terrainId ~= paint.id then
    local startPos = Vector:createVector3f(pos.x - paint.brushSize, pos.y + paint.height, pos.z - paint.brushSize);
    local endPos = Vector:createVector3f(pos.x + paint.brushSize, pos.y - 3, pos.z + paint.brushSize);

    fillTerrainGlobal(startPos, endPos, paint.id);
    --fillTerrainRadiusGlobal(pos, size, id);  -- creates a bubble/hill... not good...
  end
end



local function onPlayerCommand(event)
  local args, flags = parseArgs(event.command);
  local cmd;

  if #args >= 1 then

    if string.lower(args[1]) == "/we" then
      -- command handled
      event:setCancel(true);

      cmd = string.lower(args[2] or "");

      if cmd == "help" then
        if checkPlayerAccess(event.player, "help") then weHelp(event, table.slice(args, 3)); end;
      elseif cmd == "select" then
        if checkPlayerAccess(event.player, "select") then weSelect(event); end;
      elseif cmd == "cancel" then
        weCancel(event); -- no player access necessary...
      elseif cmd == "clear" then
        if checkPlayerAccess(event.player, "clear") then weClear(event, table.slice(args, 3), flags); end;
      elseif cmd == "fill" then
        if checkPlayerAccess(event.player, "fill") then weFill(event, table.slice(args, 3), flags); end;
      elseif cmd == "paint" then
        if checkPlayerAccess(event.player, "paint") then wePaint(event, table.slice(args, 3), flags); end;
      elseif cmd == "place" then
        if checkPlayerAccess(event.player, "place") then wePlaceBlock(event, table.slice(args, 3), flags); end;
      elseif cmd == "plant" then
        if checkPlayerAccess(event.player, "plant") then wePlant(event, table.slice(args, 3), flags); end;
      elseif cmd == "about" then
        if checkPlayerAccess(event.player, "about") then weAbout(event); end;
      else
        event.player:sendTextMessage("[#FF0000]"..i18n.t(event.player, "cmd.unknown"));
      end
    end
  end
end
addEvent("PlayerCommand", onPlayerCommand);


local function onPlayerChangePosition(event)
  if event.player:getAttribute("painting") then
    wePaintProcess(event);
  end
end
addEvent("PlayerChangePosition", onPlayerChangePosition);